### PR TITLE
prov/shm: always use CMA when sending to self

### PR DESCRIPTION
**Pull request by rwespetal**
CMA is always allowed to the same thread so use CMA when sending to self. This fixes a hang when a SHM endpoint has no peers but tries to communicate with itself.

Test description: ran fabtests and the failing test (Intel MPI osu_alltoall -n 37 -ppn 36) with and without restricted ptrace.

### CODE REVIEW

**Review by aingerson — APPROVED** [review]
The CMA enabled patch is fine but could you take out the assert? The SAR path is also used for device transfers even if CMA is available.

**Comment by rwespetal** [comment]
> The CMA enabled patch is fine but could you take out the assert? The SAR path is also used for device transfers even if CMA is available.

Ah, sorry about that. I removed that patch.

**Comment by aingerson** [comment]
Thank you!

**Comment by rwespetal** [comment]
Although I guess the same issue would apply for device transfers when sending to self and there are no other peers in the AV. Does that mean that we need to always call the SAR progress function? Or maybe set a flag on the endpoint when SAR is invoked that can be checked in the progress loop?

**Comment by aingerson** [comment]
@rwespetal I'm not sure I follow where the hang is coming from in the first place. The change makes sense to me as an optimization but I would think that an endpoint can still send to itself. 
The test case you mention (osu alltoall n 37 ppn 36) doesn't hang for me.

**Comment by aingerson** [comment]
Nevermind, I tested with CMA on. I'll retest and think about this more

**Comment by rwespetal** [comment]
Even with CMA on the test hangs. The CMA check doesn't run until another SHM peer besides itself inserted into the AV, so `SMR_CMA_CAP_NA` is still set and the progress function skips progressing SAR messages.

**Comment by aingerson** [comment]
Hmm, I think we're seeing entirely different behaviors which makes it hard to follow. I can't seem to reproduce it with CMA, only when I force it CMA_OFF. When I force to CMA_NA, it passes.
If the issue is not progressing if the cap is NA, does changing the sar_list progress condition to if cap != ON fix the issue you're seeing?
https://github.com/ofiwg/libfabric/blob/master/prov/shm/src/smr_progress.c#L950

**Comment by aingerson** [comment]
Sorry, I tested incorrectly again 🤦 
I do see the hang if I force NA or OFF and always progress the sar list does not fix the issue.
I think the problem is somewhere else. I'll keep debugging.

**Comment by rwespetal** [comment]
Huh, that's surprising. I made the change you suggested when originally debugging this (changed the check to `!SMR_CMA_CAP_OFF`)  and the test passed using SAR to do an RMA read to self.

**Comment by aingerson** [comment]
Then what the heck am I seeing...?
I guess always progressing the sar list is a one proper fix (this could actually explain a hang I was seeing with SAR device transfers...) so if you could open a PR for that, that would be great (or I can). I'll keep looking into whatever is happening on my side and see if I can figure out what my issue is.
Thanks for catching this!

**Comment by rwespetal** [comment]
Sure, I can do that. I'll retest this too to make sure I'm not misremembering anything.

**Comment by aingerson** [comment]
Thanks!
